File: Game-Dev-Math/Physics/Separating-Axis-Theorem/sat_collision.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>
#include <limits>
#include <cassert>

using namespace std;

struct Vector {
    double x, y;
    
    Vector(double _x, double _y) : x(_x), y(_y) {}
    
    double dot(const Vector& other) const {
        return x * other.x + y * other.y;
    }
    
    Vector normalize() const {
        double length = sqrt(x * x + y * y);
        if (length == 0) return Vector(0, 0);
        return Vector(x / length, y / length);
    }
};

struct Polygon {
    vector<Vector> vertices;
    Polygon(vector<Vector> v) : vertices(v) {}
};

class SeparatingAxisTheorem {
private:
    static vector<Vector> getAxes(const Polygon& poly) {
        vector<Vector> axes;
        int n = poly.vertices.size();
        for (int i = 0; i < n; i++) {
            Vector p1 = poly.vertices[i];
            Vector p2 = poly.vertices[(i + 1) % n];
            
            Vector edge(p2.x - p1.x, p2.y - p1.y);
            Vector normal(-edge.y, edge.x);
            axes.push_back(normal.normalize());
        }
        return axes;
    }

    static pair<double, double> project(const Polygon& poly, const Vector& axis) {
        double min_proj = numeric_limits<double>::infinity();
        double max_proj = -numeric_limits<double>::infinity();
        
        for (const auto& vertex : poly.vertices) {
            double projection = vertex.dot(axis);
            min_proj = min(min_proj, projection);
            max_proj = max(max_proj, projection);
        }
        return {min_proj, max_proj};
    }

public:
    static bool checkCollision(const Polygon& poly1, const Polygon& poly2) {
        vector<Vector> axes = getAxes(poly1);
        vector<Vector> axes2 = getAxes(poly2);
        axes.insert(axes.end(), axes2.begin(), axes2.end());
        
        for (const auto& axis : axes) {
            pair<double, double> p1 = project(poly1, axis);
            pair<double, double> p2 = project(poly2, axis);
            
            if (p1.second < p2.first || p2.second < p1.first) {
                return false; // Gap found, no collision
            }
        }
        return true;
    }
};

// --- CI/CD Automated Test ---
int main() {
    Polygon square1({{1, 1}, {3, 1}, {3, 3}, {1, 3}});
    Polygon square2Colliding({{2.5, 1}, {4.5, 1}, {4.5, 3}, {2.5, 3}});
    Polygon square3Safe({{5, 5}, {7, 5}, {7, 7}, {5, 7}});

    assert(SeparatingAxisTheorem::checkCollision(square1, square2Colliding) == true);
    assert(SeparatingAxisTheorem::checkCollision(square1, square3Safe) == false);

    cout << "C++ Separating Axis Theorem (SAT) Test Passed! Flawless Physics Engine Collision Verified.\n";
    return 0;
}

File: Game-Dev-Math/Physics/Separating-Axis-Theorem/satCollision.ts
class Vec2 {
    constructor(public x: number, public y: number) {}

    public dot(other: Vec2): number {
        return this.x * other.x + this.y * other.y;
    }

    public normalize(): Vec2 {
        const length = Math.sqrt(this.x * this.x + this.y * this.y);
        if (length === 0) return new Vec2(0, 0);
        return new Vec2(this.x / length, this.y / length);
    }
}

class Polygon {
    constructor(public vertices: Vec2[]) {}
}

export class SeparatingAxisTheorem {
    private static getAxes(poly: Polygon): Vec2[] {
        const axes: Vec2[] = [];
        const len = poly.vertices.length;

        for (let i = 0; i < len; i++) {
            const p1 = poly.vertices[i];
            const p2 = poly.vertices[(i + 1) % len];

            const edge = new Vec2(p2.x - p1.x, p2.y - p1.y);
            const normal = new Vec2(-edge.y, edge.x);
            axes.push(normal.normalize());
        }
        return axes;
    }

    private static project(poly: Polygon, axis: Vec2): { min: number, max: number } {
        let minProj = Infinity;
        let maxProj = -Infinity;

        for (const vertex of poly.vertices) {
            const projection = vertex.dot(axis);
            if (projection < minProj) minProj = projection;
            if (projection > maxProj) maxProj = projection;
        }

        return { min: minProj, max: maxProj };
    }

    public static checkCollision(poly1: Polygon, poly2: Polygon): boolean {
        const axes = [...this.getAxes(poly1), ...this.getAxes(poly2)];

        for (const axis of axes) {
            const p1 = this.project(poly1, axis);
            const p2 = this.project(poly2, axis);

            if (p1.max < p2.min || p2.max < p1.min) {
                // Gap found
                return false;
            }
        }

        return true;
    }
}

// --- CI/CD Automated Test ---
const square1 = new Polygon([
    new Vec2(1, 1), new Vec2(3, 1), 
    new Vec2(3, 3), new Vec2(1, 3)
]);

const square2Colliding = new Polygon([
    new Vec2(2.5, 1), new Vec2(4.5, 1), 
    new Vec2(4.5, 3), new Vec2(2.5, 3)
]);

const square3Safe = new Polygon([
    new Vec2(5, 5), new Vec2(7, 5), 
    new Vec2(7, 7), new Vec2(5, 7)
]);

const p1 = SeparatingAxisTheorem.checkCollision(square1, square2Colliding) === true;
const p2 = SeparatingAxisTheorem.checkCollision(square1, square3Safe) === false;

if (p1 && p2) {
    console.log("TypeScript Separating Axis Theorem (SAT) Test Passed!");
} else {
    process.exit(1);
}